grpc.go

```diff
@@ -623,6 +627,91 @@ func (a *agentGRPC) WaitProcess(ctx context.Context, req *pb.WaitProcessRequest)
 	}, nil
 }
 
+func getPIDIndex(title string) int {
+	// looking for PID field in ps title
+	fields := strings.Fields(title)
+	for i, f := range fields {
+		if f == "PID" {
+			return i
+		}
+	}
+	return -1
+}
+
+func (a *agentGRPC) ListProcesses(ctx context.Context, req *pb.ListProcessesRequest) (*pb.ListProcessesResponse, error) {
+	resp := &pb.ListProcessesResponse{}
+
+	c, err := a.sandbox.getContainer(req.ContainerId)
+	if err != nil {
+		return resp, err
+	}
+
+	pids, err := c.container.Processes()
+	if err != nil {
+		return resp, err
+	}
+
+	switch req.Format {
+	case "table":
+	case "json":
+		resp.ProcessList, err = json.Marshal(pids)
+		return resp, err
+	default:
+		return resp, fmt.Errorf("invalid format option")
+	}
+
+	psArgs := req.Args
+	if len(psArgs) == 0 {
+		psArgs = []string{"-ef"}
+	}
+
+	cmd := exec.Command("ps", psArgs...)
+	output, err := a.sandbox.subreaper.combinedOutput(cmd)
+	if err != nil {
+		return nil, fmt.Errorf("%s: %s", err, output)
+	}
+
+	lines := strings.Split(string(output), "\n")
+
+	pidIndex := getPIDIndex(lines[0])
+
+	// PID field not found
+	if pidIndex == -1 {
+		return nil, fmt.Errorf("failed to find PID field in ps output")
+	}
+
+	// append title
+	var result bytes.Buffer
+
+	result.WriteString(lines[0] + "\n")
+
+	for _, line := range lines[1:] {
+		if len(line) == 0 {
+			continue
+		}
+		fields := strings.Fields(line)
+		if pidIndex >= len(fields) {
+			return nil, fmt.Errorf("missing PID field: %s", line)
+		}
+
+		p, err := strconv.Atoi(fields[pidIndex])
+		if err != nil {
+			return nil, fmt.Errorf("failed to convert pid to int: %s", fields[pidIndex])
+		}
+
+		// appends pid line
+		for _, pid := range pids {
+			if pid == p {
```

Comment:
You are comparing the pid from the container namespace to the one in the agent pid namespace. Can you explain why you do this? I would imagine not processes from the container pid namespace would not be visible in the agent namespace.

Reply:
I added more comments in the code, like this

```go
// Get the list of processes that are running inside the containers.
// the PIDs match with the system PIDs, not with container's namespace
pids, err := c.container.Processes() 
```

Reply:
The fact that you are able to match the pids from the agent pid namespace makes me think that this is returning the system pids.

Reply:
sorry I don't understand your concerns , is the agent running inside a namespace? because I tested this patch and works, `ps` returns all the processes, even I see the PID 1 (systemd)

Reply:
I think that is sort of the crux - when you do a 'ps' of a container, does it show you the systemd process '1' of the (for want of a better phrase..) mini-OS that is in the VM. If I am understanding correctly, a `ps` of a container should only show you the processes in that container - in that workload - we should not see anything, such as systemd pid1, from the mini-OS.

Maybe you can post an example output of what happens to for a 'busybox sh' container or similar to clarify what we get? Also, maybe, a `ps` from inside the container (`docker exec ps`) so we can compare?

Reply:
with runc

```
$ docker run --runtime=runc -dti debian bash
d7d8f6ed73faef744b3961efdef780a4b94642b6d396bd73afa5699c32ee7d01
$ sudo docker-runc ps d7d8f6ed73faef744b3961efdef780a4b94642b6d396bd73afa5699c32ee7d01
UID        PID  PPID  C STIME TTY          TIME CMD
root      4337  4320  0 08:31 pts/0    00:00:00 bash
```

with kata-runtime

```
$ docker run --runtime=kata-runtime -dti debian bash
32b6d8eb33c750c8ebf960464a0a40860e138d24bfd13a45ee45704dbd1462bc
$ sudo kata-runtime ps 32b6d8eb33c750c8ebf960464a0a40860e138d24bfd13a45ee45704dbd1462bc
UID         PID   PPID  C STIME TTY          TIME CMD
root        147    115  0 13:44 ?        00:00:00 bash
```

Reply:
docker exec with runc

```
$ docker run --runtime=runc -dti busybox sh
5e93234a965d7e64d7a2ab918b9eaace72712137f08b6a43ce61162f45752ad0
$ docker exec 5e93234a965d7e64d7a2ab918b9eaace72712137f08b6a43ce61162f45752ad0 ps
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    6 root      0:00 ps
```


docker exec with kata-runtime
```
$ docker run --runtime=kata-runtime -dti busybox sh
9e1070e509f5574117e86319aff9c1c2d64a00f2d27a40b974c46f59971266c3
$ docker exec 9e1070e509f5574117e86319aff9c1c2d64a00f2d27a40b974c46f59971266c3 ps
PID   USER     TIME  COMMAND
    1 root      0:00 sh
    5 root      0:00 ps
```

Reply:
Ooh. I see. That is interesting - that it seems `docker ps` does not use the PIDS from inside the container user namepace, but indeed shows the host space PIDS. Not really what I expected at all (but maybe there are sound reasons for `runc` - that maybe you can then manipulate those processes from the guest).
Things are a little different for kata, as even with that 'guest PID', that is really the VM root PID, so there is not really much you can do with that from the guest - but, showing that is probably as near to `runc` behaviour as we can get. We could debate if that is meaningful for VM containers though.
@amshinde , you agree here that this is fine then?

Reply:
Yeah, I was not expecting to see the system processes here. Looks fine to me then.